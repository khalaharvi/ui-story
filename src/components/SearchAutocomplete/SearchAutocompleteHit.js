import React from 'react';
import PropTypes from 'prop-types';

import { cx } from './SearchAutocomplete';

const SearchAutocompleteHit = ({ hit, hitTitleKey, hitImageKey }) => {
  const titleKey = typeof hitTitleKey === 'string' ? hitTitleKey : hitTitleKey.find(key => !!hit[key]);
  let hitImg = null;

  if (hitImageKey) {
    const imgKey = (typeof hitImageKey === 'string') ? hitImageKey : hitImageKey.find(key => !!hit[key]);

    hitImg = imgKey && hit[imgKey] ?
      <div className={cx('hit-image')}><img src={hit[imgKey]} alt={hit[titleKey]} /></div> :
      <div className={cx('hit-image')}><img src="http://via.placeholder.com/40" alt="placeholder" /></div>;
  }

  return (
    <div className={cx('hit')}>{hitImg}<span className={cx('hit-title')}>{hit[titleKey]}</span></div>
  );
};

SearchAutocompleteHit.propTypes = {
  hit: PropTypes.object.isRequired, // eslint-disable-line react/forbid-prop-types
  hitTitleKey: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.arrayOf(PropTypes.string)
  ]).isRequired,
  hitImageKey: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.arrayOf(PropTypes.string)
  ])
};

export default SearchAutocompleteHit;
