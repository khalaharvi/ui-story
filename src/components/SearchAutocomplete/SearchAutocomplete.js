import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { connectAutoComplete } from 'react-instantsearch/connectors';
import Autosuggest from 'react-autosuggest';

import { createClassNames } from 'utils';
import SearchAutocompleteHit from './SearchAutocompleteHit';
import SearchAutocompleteInput from './SearchAutocompleteInput';

export const cx = createClassNames('search-autocomplete');

export class SearchAutocomplete extends Component {
  state = {
    value: this.props.currentRefinement
  };

  onSuggestionsClearRequested = () => {
    const { refine } = this.props;

    /**
     * Workaround to be able to get the input value when hitting the submit button
     * Current input value is stored before it gets cleared by "refine('')"
     */
    if (this.input) {
      this.setState({ value: this.input.value });
    }

    refine('');
  };

  onSuggestionSelected = (event, {
    suggestion,
    suggestionValue,
    suggestionIndex,
    sectionIndex,
    method
  }) => {
    if (method === 'enter') {
      event.preventDefault();
    }

    const { onHitSelected } = this.props;

    if (onHitSelected) {
      onHitSelected(event, {
        hit: suggestion,
        hitValue: suggestionValue,
        hitIndex: suggestionIndex,
        sectionIndex,
        method
      });
    }
  };

  onInputChange = (e) => {
    const { onChange } = this.props;

    this.setState({ value: e.target.value });

    if (onChange) {
      onChange(e);
    }
  };

  onFormSubmit = (e) => {
    e.preventDefault();

    const { onSubmit } = this.props;

    if (this.input) {
      this.input.blur();
    }

    if (onSubmit) {
      onSubmit(this.state.value);
    }

    this.setState({ value: '' });
  };

  render() {
    const {
      // connectAutoComplete props
      currentRefinement,
      hits,
      refine,

      className,
      getHitValue,
      highlightFirstHit,
      hitImageKey,
      hitTitleKey,
      multiSection,
      placeholder,
      renderHit,
      renderSectionTitle,
      inputRef,
      transformHit
    } = this.props;

    const renderSuggestion = renderHit || (hit => (
      <SearchAutocompleteHit
        hit={transformHit ? transformHit(hit) : hit}
        hitTitleKey={hitTitleKey}
        hitImageKey={hitImageKey}
      />
    ));

    const reactAutosuggestClassnames = {
      container: classnames(cx('container'), className),
      containerOpen: cx('container--open'),
      input: cx('input'),
      inputOpen: cx('input--open'),
      inputFocused: cx('input--focused'),
      suggestionsContainer: cx('suggestions-container'),
      suggestionsContainerOpen: cx('suggestions-container--open'),
      suggestionsList: cx('suggestions-list'),
      suggestion: cx('suggestion'),
      suggestionFirst: cx('suggestion--first'),
      suggestionHighlighted: cx('suggestion--highlighted'),
      sectionContainer: cx('section-container'),
      sectionContainerFirst: cx('section-container--first'),
      sectionTitle: cx('section-title')
    };

    return (
      <form onSubmit={this.onFormSubmit} noValidate action="" role="search">
        <Autosuggest
          ref={(autosuggest) => {
            if (autosuggest !== null) {
              this.input = autosuggest.input;

              if (inputRef) {
                inputRef(autosuggest.input);
              }
            }
          }}
          getSectionSuggestions={section => section.hits}
          getSuggestionValue={getHitValue}
          highlightFirstSuggestion={highlightFirstHit}
          inputProps={{
            onChange: this.onInputChange,
            placeholder,
            value: currentRefinement
          }}
          multiSection={multiSection}
          onSuggestionsClearRequested={this.onSuggestionsClearRequested}
          onSuggestionSelected={this.onSuggestionSelected}
          onSuggestionsFetchRequested={({ value }) => refine(value)}
          renderInputComponent={SearchAutocompleteInput}
          renderSectionTitle={renderSectionTitle}
          renderSuggestion={renderSuggestion}
          suggestions={hits}
          theme={reactAutosuggestClassnames}
          focusInputOnSuggestionClick={false}
        />
      </form>
    );
  }
}

SearchAutocomplete.propTypes = {
  currentRefinement: PropTypes.string,
  hits: PropTypes.array.isRequired, // eslint-disable-line react/forbid-prop-types
  refine: PropTypes.func.isRequired,

  className: PropTypes.string,
  getHitValue: PropTypes.func.isRequired,
  highlightFirstHit: PropTypes.bool,
  hitImageKey: PropTypes.string,
  hitTitleKey: (props, propName) => {
    const hitTitleKey = props[propName];

    if (props.renderHit === undefined && typeof hitTitleKey !== 'string') {
      throw new Error("'hitTitleKey' is required.");
    }
  },
  multiSection: PropTypes.bool,
  onChange: PropTypes.func,
  onSubmit: PropTypes.func,
  onHitSelected: PropTypes.func,
  placeholder: PropTypes.string,
  renderHit: PropTypes.func,
  renderSectionTitle: (props, propName) => {
    const renderSectionTitle = props[propName];

    if (props.multiSection === true && typeof renderSectionTitle !== 'function') {
      throw new Error("'renderSectionTitle' is required.");
    }
  },
  inputRef: PropTypes.func,
  transformHit: PropTypes.func
};

SearchAutocomplete.defaultProps = {
  currentRefinement: '',
  highlightFirstHit: false,
  multiSection: false,
  placeholder: 'Search...',
  onChange: () => {}
};

export default connectAutoComplete(SearchAutocomplete);
