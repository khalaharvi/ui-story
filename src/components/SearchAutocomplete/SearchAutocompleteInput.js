import React from 'react';

import { Search as SearchIcon } from 'icons';
import { cx } from './SearchAutocomplete';

const SearchAutocompleteInput = props => (
  <div className={cx('input-container')}>
    <input {...props} type="search" />
    <button className={cx('input-button')} type="submit">
      <SearchIcon className={cx('input-icon')} />
    </button>
  </div>
);

export default SearchAutocompleteInput;
