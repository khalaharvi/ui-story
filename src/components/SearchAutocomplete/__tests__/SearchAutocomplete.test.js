import React from 'react';
import { mount, shallow } from 'enzyme';

import { SearchAutocomplete } from '../SearchAutocomplete';

const getSuggestionValue = hit => hit.title;
const hitTitleKey = 'title';

describe('<SearchAutocomplete />', () => {
  it('renders correctly', () => {
    const wrapper = shallow(<SearchAutocomplete
      hits={[]}
      refine={() => null}
      getHitValue={getSuggestionValue}
      hitTitleKey={hitTitleKey}
    />);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders correctly with className', () => {
    const wrapper = shallow(<SearchAutocomplete
      hits={[]}
      refine={() => null}
      getHitValue={getSuggestionValue}
      hitTitleKey={hitTitleKey}
      className="custom-search-autocomplete"
    />);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders correctly with a custom placeholder', () => {
    const wrapper = shallow(<SearchAutocomplete
      hits={[]}
      refine={() => null}
      placeholder="Search products..."
      getHitValue={getSuggestionValue}
      hitTitleKey={hitTitleKey}
    />);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders correctly with transformHit', () => {
    const wrapper = shallow(<SearchAutocomplete
      hits={[]}
      refine={() => null}
      getHitValue={getSuggestionValue}
      hitTitleKey={hitTitleKey}
      transformHit={hit => hit}
    />);

    expect(wrapper).toMatchSnapshot();
  });

  it('treats its currentRefinement prop as its input value', () => {
    const wrapper = shallow(<SearchAutocomplete
      hits={[]}
      refine={() => null}
      currentRefinement="product"
      getHitValue={getSuggestionValue}
      hitTitleKey={hitTitleKey}
    />);

    expect(wrapper).toMatchSnapshot();
  });

  it('refines its value on change', () => {
    const refine = jest.fn();
    const wrapper = mount(<SearchAutocomplete
      hits={[]}
      refine={refine}
      getHitValue={getSuggestionValue}
      hitTitleKey={hitTitleKey}
    />);

    wrapper.find('input').simulate('change', { target: { value: 'chair' } });

    expect(refine.mock.calls).toHaveLength(1);
    expect(refine.mock.calls[0][0]).toBe('chair');

    wrapper.unmount();
  });

  it('refines its value on blur', () => {
    const refine = jest.fn();
    const wrapper = mount(<SearchAutocomplete
      hits={[]}
      refine={refine}
      getHitValue={getSuggestionValue}
      hitTitleKey={hitTitleKey}
    />);

    wrapper.find('input').simulate('blur');

    expect(refine.mock.calls).toHaveLength(1);
    expect(refine.mock.calls[0][0]).toBe('');

    wrapper.unmount();
  });

  it('renders correctly with onHitSelected', () => {
    const wrapper = shallow(<SearchAutocomplete
      hits={[]}
      refine={() => null}
      getHitValue={getSuggestionValue}
      hitTitleKey={hitTitleKey}
      onHitSelected={() => null}
    />);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders correctly with onSubmit', () => {
    const wrapper = shallow(<SearchAutocomplete
      hits={[]}
      refine={() => null}
      getHitValue={getSuggestionValue}
      hitTitleKey={hitTitleKey}
      onSubmit={() => null}
    />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should call onSubmit with current input value', () => {
    const onSubmit = jest.fn();
    const wrapper = mount(<SearchAutocomplete
      hits={[]}
      refine={() => null}
      getHitValue={getSuggestionValue}
      hitTitleKey={hitTitleKey}
      onSubmit={onSubmit}
    />);

    wrapper.find('input').simulate('change', { target: { value: 'chair' } });
    wrapper.find('form').simulate('submit');

    expect(onSubmit.mock.calls).toHaveLength(1);
    expect(onSubmit.mock.calls[0][0]).toBe('chair');

    wrapper.find('input').simulate('change', { target: { value: 'lamp' } });
    wrapper.find('[type="submit"]').simulate('submit');

    expect(onSubmit.mock.calls).toHaveLength(2);
    expect(onSubmit.mock.calls[1][0]).toBe('lamp');

    wrapper.unmount();
  });
});
