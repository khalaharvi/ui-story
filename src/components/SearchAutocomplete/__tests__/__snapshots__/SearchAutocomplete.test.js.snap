// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchAutocomplete /> renders correctly 1`] = `
<form
  action=""
  noValidate={true}
  onSubmit={[Function]}
  role="search"
>
  <Autosuggest
    alwaysRenderSuggestions={false}
    focusInputOnSuggestionClick={false}
    getSectionSuggestions={[Function]}
    getSuggestionValue={[Function]}
    highlightFirstSuggestion={false}
    id="1"
    inputProps={
      Object {
        "onChange": [Function],
        "placeholder": "Search...",
        "value": "",
      }
    }
    multiSection={false}
    onSuggestionSelected={[Function]}
    onSuggestionsClearRequested={[Function]}
    onSuggestionsFetchRequested={[Function]}
    renderInputComponent={[Function]}
    renderSuggestion={[Function]}
    renderSuggestionsContainer={[Function]}
    shouldRenderSuggestions={[Function]}
    suggestions={Array []}
  />
</form>
`;

exports[`<SearchAutocomplete /> renders correctly with a custom placeholder 1`] = `
<form
  action=""
  noValidate={true}
  onSubmit={[Function]}
  role="search"
>
  <Autosuggest
    alwaysRenderSuggestions={false}
    focusInputOnSuggestionClick={false}
    getSectionSuggestions={[Function]}
    getSuggestionValue={[Function]}
    highlightFirstSuggestion={false}
    id="1"
    inputProps={
      Object {
        "onChange": [Function],
        "placeholder": "Search products...",
        "value": "",
      }
    }
    multiSection={false}
    onSuggestionSelected={[Function]}
    onSuggestionsClearRequested={[Function]}
    onSuggestionsFetchRequested={[Function]}
    renderInputComponent={[Function]}
    renderSuggestion={[Function]}
    renderSuggestionsContainer={[Function]}
    shouldRenderSuggestions={[Function]}
    suggestions={Array []}
  />
</form>
`;

exports[`<SearchAutocomplete /> renders correctly with className 1`] = `
<form
  action=""
  noValidate={true}
  onSubmit={[Function]}
  role="search"
>
  <Autosuggest
    alwaysRenderSuggestions={false}
    focusInputOnSuggestionClick={false}
    getSectionSuggestions={[Function]}
    getSuggestionValue={[Function]}
    highlightFirstSuggestion={false}
    id="1"
    inputProps={
      Object {
        "onChange": [Function],
        "placeholder": "Search...",
        "value": "",
      }
    }
    multiSection={false}
    onSuggestionSelected={[Function]}
    onSuggestionsClearRequested={[Function]}
    onSuggestionsFetchRequested={[Function]}
    renderInputComponent={[Function]}
    renderSuggestion={[Function]}
    renderSuggestionsContainer={[Function]}
    shouldRenderSuggestions={[Function]}
    suggestions={Array []}
  />
</form>
`;

exports[`<SearchAutocomplete /> renders correctly with onHitSelected 1`] = `
<form
  action=""
  noValidate={true}
  onSubmit={[Function]}
  role="search"
>
  <Autosuggest
    alwaysRenderSuggestions={false}
    focusInputOnSuggestionClick={false}
    getSectionSuggestions={[Function]}
    getSuggestionValue={[Function]}
    highlightFirstSuggestion={false}
    id="1"
    inputProps={
      Object {
        "onChange": [Function],
        "placeholder": "Search...",
        "value": "",
      }
    }
    multiSection={false}
    onSuggestionSelected={[Function]}
    onSuggestionsClearRequested={[Function]}
    onSuggestionsFetchRequested={[Function]}
    renderInputComponent={[Function]}
    renderSuggestion={[Function]}
    renderSuggestionsContainer={[Function]}
    shouldRenderSuggestions={[Function]}
    suggestions={Array []}
  />
</form>
`;

exports[`<SearchAutocomplete /> renders correctly with onSubmit 1`] = `
<form
  action=""
  noValidate={true}
  onSubmit={[Function]}
  role="search"
>
  <Autosuggest
    alwaysRenderSuggestions={false}
    focusInputOnSuggestionClick={false}
    getSectionSuggestions={[Function]}
    getSuggestionValue={[Function]}
    highlightFirstSuggestion={false}
    id="1"
    inputProps={
      Object {
        "onChange": [Function],
        "placeholder": "Search...",
        "value": "",
      }
    }
    multiSection={false}
    onSuggestionSelected={[Function]}
    onSuggestionsClearRequested={[Function]}
    onSuggestionsFetchRequested={[Function]}
    renderInputComponent={[Function]}
    renderSuggestion={[Function]}
    renderSuggestionsContainer={[Function]}
    shouldRenderSuggestions={[Function]}
    suggestions={Array []}
  />
</form>
`;

exports[`<SearchAutocomplete /> renders correctly with transformHit 1`] = `
<form
  action=""
  noValidate={true}
  onSubmit={[Function]}
  role="search"
>
  <Autosuggest
    alwaysRenderSuggestions={false}
    focusInputOnSuggestionClick={false}
    getSectionSuggestions={[Function]}
    getSuggestionValue={[Function]}
    highlightFirstSuggestion={false}
    id="1"
    inputProps={
      Object {
        "onChange": [Function],
        "placeholder": "Search...",
        "value": "",
      }
    }
    multiSection={false}
    onSuggestionSelected={[Function]}
    onSuggestionsClearRequested={[Function]}
    onSuggestionsFetchRequested={[Function]}
    renderInputComponent={[Function]}
    renderSuggestion={[Function]}
    renderSuggestionsContainer={[Function]}
    shouldRenderSuggestions={[Function]}
    suggestions={Array []}
  />
</form>
`;

exports[`<SearchAutocomplete /> treats its currentRefinement prop as its input value 1`] = `
<form
  action=""
  noValidate={true}
  onSubmit={[Function]}
  role="search"
>
  <Autosuggest
    alwaysRenderSuggestions={false}
    focusInputOnSuggestionClick={false}
    getSectionSuggestions={[Function]}
    getSuggestionValue={[Function]}
    highlightFirstSuggestion={false}
    id="1"
    inputProps={
      Object {
        "onChange": [Function],
        "placeholder": "Search...",
        "value": "product",
      }
    }
    multiSection={false}
    onSuggestionSelected={[Function]}
    onSuggestionsClearRequested={[Function]}
    onSuggestionsFetchRequested={[Function]}
    renderInputComponent={[Function]}
    renderSuggestion={[Function]}
    renderSuggestionsContainer={[Function]}
    shouldRenderSuggestions={[Function]}
    suggestions={Array []}
  />
</form>
`;
